common/lib/common-utils/src/idleTaskScheduler.ts

```diff
@@ -0,0 +1,67 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryProperties } from "@fluidframework/common-definitions";
+
+
+interface Task {
+    handle: () => any,
+}
+
+
+/**
+ * Helper class that is used to schedule non-essential tasks
+ * Time measurements are in milliseconds as a floating point with a decimal
+ */
+export default class IdleTaskScheduler {
+    //List of tasks waiting to be run
+    private idleTaskList:  Task[] = [];
+    // A reference to the task currently being processed.
+    private taskHandle = 0;
+
+    /*
+    Add tasks to FIFO queue of tasks that are run during idle callback period.
+    Creates idle callback if it doesn't exist with a timeout of 1s.
+    */
+    public enqueueTask<T>(task: () => T){
+        let newTask: Task;
+        newTask = {
+            handle: task
+        }
+        this.idleTaskList.push(newTask)
+
+        if (this.taskHandle) {
+            this.taskHandle = requestIdleCallback(this.runTaskQueue, { timeout: 1000 });
```

Comment:
I'd think requestIdleCallback does not exist in node.js environment. Does it?
If it does not, we need some equivalence or leverage of promises instead.

Reply:
Somewhat similar question - does it exists in web worker? Ideally our code can run in any environment (i.e. be isomorphic)

Reply:
If it exists only on the web, consider using window.requestIdleCallback and testing that such thing exists, i.e.
```ts
if (typeof window === "object" && typeof window?.requestIdleCallback === "function") {
    // use funcitonality
} else {
    // some fallback
}
```